dissect/target/plugins/os/windows/notifications.py

```diff
@@ -55,19 +192,172 @@ def __init__(self, target):
+                    type=tile.Type,
+                    index=tile.Index,
+                    name=name,
+                    content=tile_xml,
+                    _target=self.target,
+                    _user=user,
+                )
+
+                tile_records.append(tile_record)
+        return tile_records
+
+    def _get_appdb_toast_records(self, chunk: c_appdb.Chunk, user: RecordDescriptor) -> list[AppDBToastRecord]:
+        toast_records = []
+        num_toasts = len(chunk.Toasts)
+
+        for toast_no in range(num_toasts):
+            toast = chunk.Toasts[toast_no]
+
+            if toast.UniqueId:
+                toast_arrival_time = None
+                if ts := toast.ArrivalTime:
+                    toast_arrival_time = wintimestamp(ts)
+                toast_expiry_time = None
+                if ts := toast.ExpiryTime:
+                    toast_expiry_time = wintimestamp(ts)
+                name1 = toast.Name1.strip("\x00")
+                name2 = toast.Name2.strip("\x00")
+
+                xml_size = toast.ContentLength
+                toast_xml = chunk.ToastXml[toast_no].Content[:xml_size]
+                toast_xml = toast_xml.decode("utf-8", errors="surrogateescape")
+
+                toast_record = AppDBToastRecord(
+                    id=toast.UniqueId,
+                    arrival_time=toast_arrival_time,
+                    expiry_time=toast_expiry_time,
+                    type=toast.Type,
+                    index=toast.Index,
+                    name1=name1,
+                    name2=name2,
+                    content=toast_xml,
+                    _target=self.target,
+                    _user=user,
+                )
+
+                toast_records.append(toast_record)
+        return toast_records
+
+    @export(record=GroupedRecord)
+    def appdb(self) -> Iterator[GroupedRecord]:
+        for user, appdb_file in self.appdb_files:
+            with appdb_file.open(mode="rb") as fp:
+                timestamp = None
+                for chunk_no in range(NUM_APPDB_CHUNKS):
+                    chunk = c_appdb.Chunk(fp)
+
+                    if chunk_no == 0:
+                        timestamp = wintimestamp(chunk.Header.Timestamp)
+
+                    if chunk.Info.InUse == 0:
+                        continue
+                    elif chunk.Info.InUse != 1:
+                        self.target.log.warning(
+                            "Unknown field value %s for chunk.Info.InUse, "
+                            "please consider providing us with a sample.",
+                            chunk.Info.InUse,
+                        )
+                        continue
+
+                    chunk_record = self._get_appdb_chunk_record(chunk, timestamp, user)
+                    tile_records = self._get_appdb_tile_records(chunk, user)
+                    toast_records = self._get_appdb_toast_records(chunk, user)
+
+                    yield GroupedRecord(
```

Comment:
I haven't checked myself so perhaps I should :smile: but is a GroupedRecord the best method? Have you considered individual records?

Reply:
I'm not sure if we can say that the tiles and toasts are independent. It is also unclear that if there are tiles and toasts, the first tile and first toast always go together etc.

We could emit the chunk, tiles and toasts individually and add the chunk id to the latter two, but then you'd have to tie everything back together again at the receiving end.

Reply:
I added a commit to normalize the records: separate push,badge, tile and toast records. This also allows the timestamps to come first. I think this makes it at least easier to answer the question what was shown to a user when.